File: examples/python/tests/test_renderer.py
"""Tests for renderer module."""

import math
import unittest

from circle import Circle
from rectangle import Rectangle
from renderer import Renderer
from shape import Color
from vector3 import Vector3


def _make_renderer() -> Renderer:
    r = Renderer()
    r.add(Circle(Vector3(0.0, 0.0, 0.0), 5.0, Color.RED))
    r.add(Rectangle(Vector3(0.0, 0.0, 0.0), 4.0, 6.0, Color.YELLOW))
    return r


class TestAddAndCount(unittest.TestCase):
    def test_empty(self):
        r = Renderer()
        self.assertEqual(r.count, 0)

    def test_after_add(self):
        r = _make_renderer()
        self.assertEqual(r.count, 2)


class TestTotalArea(unittest.TestCase):
    def test_computed(self):
        r = _make_renderer()
        expected = math.pi * 25.0 + 24.0
        self.assertAlmostEqual(r.total_area(), expected)


class TestFilter(unittest.TestCase):
    def test_by_area(self):
        r = _make_renderer()
        big = r.filter(lambda s: s.area() > 25)
        self.assertEqual(len(big), 1)
        self.assertEqual(big[0].name, "Circle")

    def test_empty_result(self):
        r = _make_renderer()
        none = r.filter(lambda s: s.area() > 1000)
        self.assertEqual(len(none), 0)


class TestSortedByArea(unittest.TestCase):
    def test_ascending(self):
        r = _make_renderer()
        shapes = r.sorted_by_area()
        self.assertLessEqual(shapes[0].area(), shapes[1].area())

    def test_descending(self):
        r = _make_renderer()
        shapes = r.sorted_by_area(reverse=True)
        self.assertGreaterEqual(shapes[0].area(), shapes[1].area())


class TestRender(unittest.TestCase):
    def test_header(self):
        r = _make_renderer()
        output = r.render()
        self.assertIn("Renderer (2 shapes)", output)

    def test_total_area_in_output(self):
        r = _make_renderer()
        output = r.render()
        self.assertIn("Total area:", output)


class TestIterShapes(unittest.TestCase):
    def test_yields_all(self):
        r = _make_renderer()
        shapes = list(r.iter_shapes())
        self.assertEqual(len(shapes), 2)

    def test_order_preserved(self):
        r = _make_renderer()
        shapes = list(r.iter_shapes())
        self.assertEqual(shapes[0].name, "Circle")
        self.assertEqual(shapes[1].name, "Rectangle")

File: examples/rust/tests/test_renderer.rs
extern crate shapes;

use shapes::circle::Circle;
use shapes::rectangle::Rectangle;
use shapes::renderer::Renderer;
use shapes::shape::{Color, Shape};
use shapes::vector3::Vec3;
use std::f64::consts::PI;

fn make_circle(r: f64) -> Box<dyn Shape> {
    Box::new(Circle::new(Vec3::new(0.0, 0.0, 0.0), r, Color::Blue))
}

fn make_rect(w: f64, h: f64) -> Box<dyn Shape> {
    Box::new(Rectangle::new(Vec3::new(0.0, 0.0, 0.0), w, h, Color::Green))
}

#[test]
fn starts_empty() {
    let r = Renderer::new();
    assert_eq!(r.count(), 0);
    assert!((r.total_area() - 0.0).abs() < 1e-10);
}

#[test]
fn counts_shapes() {
    let mut r = Renderer::new();
    r.add(make_circle(1.0));
    r.add(make_rect(2.0, 3.0));
    assert_eq!(r.count(), 2);
}

#[test]
fn computes_total_area() {
    let mut r = Renderer::new();
    r.add(make_circle(1.0));
    r.add(make_rect(2.0, 3.0));
    let expected = PI * 1.0 + 6.0;
    assert!((r.total_area() - expected).abs() < 1e-10);
}

#[test]
fn filters_by_color() {
    let mut r = Renderer::new();
    r.add(make_circle(1.0));   // Blue
    r.add(make_rect(2.0, 3.0)); // Green
    r.add(make_circle(2.0));   // Blue
    let blues = r.filter(|s| s.color() == Color::Blue);
    assert_eq!(blues.len(), 2);
}

#[test]
fn filters_by_area() {
    let mut r = Renderer::new();
    r.add(make_rect(3.0, 3.0));   // 9
    r.add(make_rect(10.0, 5.0));  // 50
    r.add(make_circle(1.0));      // ~3.14
    let big = r.filter(|s| s.area() > 20.0);
    assert_eq!(big.len(), 1);
}

#[test]
fn renders_to_writer() {
    let mut r = Renderer::new();
    r.add(make_circle(1.0));
    let mut buf = Vec::new();
    r.render(&mut buf);
    let output = String::from_utf8(buf).unwrap();
    assert!(output.contains("Rendering 1 shape(s)"));
    assert!(output.contains("Circle"));
}
